nanoFramework.Device.Bluetooth/BluetoothLEServer.cs

```diff
@@ -0,0 +1,267 @@
+        /// </summary>
+        /// <remarks>
+        /// The primary service will be index 0 followed by the Device Information at index 1.
+        /// Any other Services added to server will follow these in the order they were created.
+        /// </remarks>
+        public GattServiceProvider[] Services { get { return (GattServiceProvider[])_services.ToArray(typeof(GattServiceProvider)); } }
+
+        #endregion
+
+        #region Security/Pairing
+
+        /// <summary>
+        /// Return Pairing object for handling Pairing.
+        /// </summary>
+        public DevicePairing Pairing { get => _pairing; }
+
+        /// <summary>
+        /// Evaulate all services and work out security requirements
+        /// </summary>
+        /// <returns>Minimum DevicePairingProtectionLevel for servcies/characteristics</returns>
+        private DevicePairingProtectionLevel EvaluateSecurityPairingRequirements()
+        {
+            DevicePairingProtectionLevel _minProtectionLevel = DevicePairingProtectionLevel.Default;
+
+            bool secureConnection = false;
+            bool authentication = false;
+
+            foreach (GattServiceProvider srvp in _services)
+            {
+                foreach (GattLocalCharacteristic chr in srvp.Service.Characteristics)
+                {
+                    if (chr.ReadProtectionLevel == GattProtectionLevel.EncryptionRequired ||
+                        chr.ReadProtectionLevel == GattProtectionLevel.EncryptionAndAuthenticationRequired ||
+                        chr.WriteProtectionLevel == GattProtectionLevel.EncryptionRequired ||
+                        chr.WriteProtectionLevel == GattProtectionLevel.EncryptionAndAuthenticationRequired)
+                    {
+                        secureConnection = true;
+                    }
+
+                    if (chr.ReadProtectionLevel == GattProtectionLevel.AuthenticationRequired ||
+                        chr.ReadProtectionLevel == GattProtectionLevel.EncryptionAndAuthenticationRequired ||
+                        chr.WriteProtectionLevel == GattProtectionLevel.AuthenticationRequired ||
+                        chr.WriteProtectionLevel == GattProtectionLevel.EncryptionAndAuthenticationRequired)
+                    {
+                        authentication = true;
+                    }
+                }
+            }
+
+            if (secureConnection)
+            {
+                _minProtectionLevel = DevicePairingProtectionLevel.Encryption;
+            }
+
+            if (authentication)
+            {
+                _minProtectionLevel = DevicePairingProtectionLevel.EncryptionAndAuthentication;
+            }
+
+            return _minProtectionLevel;
+        }
+
+        #endregion Security
+
+        /// <summary>
+        /// Start Bluetoth stack for server mode
+        /// If already running or in Client mode will give an InvalidOperation Exception.
```

Comment:
You should use the /// comment with the exception pattern. Makes things easier to read from a developer point of view.